pallets/allocations/src/lib.rs

```diff
@@ -65,6 +65,74 @@ impl Default for Releases {
 	}
 }
 
+#[derive(Default, TypeInfo)]
+pub struct MintCurve<T: Config> {
+	session_period: T::BlockNumber,
+	fiscal_period: T::BlockNumber,
+	inflation_steps: Vec<Perbill>,
+	maximum_supply: BalanceOf<T>,
+}
+
+impl<T: Config> MintCurve<T> {
+	pub fn new(
+		session_period: T::BlockNumber,
+		fiscal_period: T::BlockNumber,
+		inflation_steps: &[Perbill],
+		maximum_supply: BalanceOf<T>,
+	) -> Self {
+		Self {
+			// Enforce a session period is at least one block
+			session_period: session_period.max(One::one()),
+			// Enforce a fiscal period is greater or equal a session period
+			fiscal_period: fiscal_period.max(session_period),
+			inflation_steps: inflation_steps.to_vec(),
+			maximum_supply,
+		}
+	}
+
+	pub fn checked_calc_next_session_quota(
+		&self,
+		block_number: T::BlockNumber,
+		current_supply: BalanceOf<T>,
+		forced: bool,
+	) -> Option<BalanceOf<T>> {
+		if (block_number.checked_rem(&self.fiscal_period) == Some(T::BlockNumber::zero())) || forced {
+			let step: usize = block_number
+				.checked_div(&self.fiscal_period)
+				.unwrap_or_else(Zero::zero)
+				.unique_saturated_into();
+			let max_inflation_rate = *self
+				.inflation_steps
+				.get(step)
+				.or_else(|| self.inflation_steps.last())
+				.unwrap_or(&Zero::zero());
+			let target_increase =
+				(self.maximum_supply.saturating_sub(current_supply)).min(max_inflation_rate * current_supply);
+			let session_quota = Perbill::from_rational(self.session_period, self.fiscal_period) * target_increase;
+			Some(session_quota)
```

Comment:
There are a few unchecked multiplications

Reply:
The type Perbill features an internal saturation promise which makes it safe to be used in that way. The following passing test may better explain it:
```
#[test]
fn perbill_mul_test() {
	assert_eq!(Perbill::from_parts(u32::MAX), Perbill::from_percent(100));
	assert_eq!(Perbill::from_perthousand(u32::MAX) * u64::MAX, u64::MAX);
}
```